s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -65,6 +65,31 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials,
     Source.fromFuture(request(s3Location, range).flatMap(entityForSuccess).map(_.dataBytes)).flatMapConcat(identity)
   }
 
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
+    import system.dispatcher
+
+    /**
+     *
+     * @param args The Option[String] is our continuation token and the Boolean is our holder
+     *             for wether this is the first time the method was called. When the token is None
+     *             and the first call Boolean is false, we know that we are done getting keys.
+     * @return
+     */
+    def listBucketCall(args: (Option[String], Boolean)): Future[Option[((Option[String], Boolean), Seq[String])]] =
```

Comment:
Ya, that would be a little cleaner for sure. Ill see what I can come up. Will update this soon.